File: tools/eds_to_yaml.py
# eds_to_yaml.py - Convert EDS to YAML format
import sys
import os
import yaml
import json
from datetime import datetime, timezone
from eds_codec import eds_to_json

SCHEMA_VERSION = "1.0.0"
TOOL_VERSION = "1.0.0"

# Custom list type for flow-style output
class FlowList(list):
    """List that will be dumped in flow style."""
    pass

def represent_flow_list(dumper, data):
    return dumper.represent_sequence('tag:yaml.org,2002:seq', data, flow_style=True)

yaml.add_representer(FlowList, represent_flow_list)

# Load defaults from schema.yaml
def load_schema():
    """Load element type defaults and version from schema.yaml."""
    global SCHEMA_VERSION
    schema_path = os.path.join(os.path.dirname(__file__), "schema.yaml")
    with open(schema_path, "r", encoding="utf-8") as f:
        schema = yaml.safe_load(f)

    # Get schema version
    SCHEMA_VERSION = schema.get("version", "1.0.0")

    defaults = {}
    for type_name, type_def in schema.get("types", {}).items():
        defaults[type_name] = type_def.get("defaults", {})
    return defaults

DEFAULTS = load_schema()

def strip_defaults(node_type, props):
    """Remove properties that match defaults, keeping only non-default values."""
    defaults = DEFAULTS.get(node_type, {})
    result = {}
    for key, value in props.items():
        if key in ("type", "nr"):
            continue  # Skip type and nr, handled separately
        # Skip auto-generated properties
        if key == "naam" and node_type == "Kring":
            continue  # Kring names are auto-generated
        if key in ("autonr", "autoKringNaam") and value == "auto":
            continue  # Skip auto values
        default_val = defaults.get(key)
        # Compare as strings since EDS often stores numbers as strings
        if str(value) != str(default_val) if default_val is not None else True:
            result[key] = value
    return result

def build_tree(data):
    """Convert flat EDS data to tree structure."""
    # Build id -> item lookup
    id_to_item = {item["id"]: item for item in data}

    # Build parent -> children lookup
    children_map = {}
    for item in data:
        parent = item["parent"]
        if parent not in children_map:
            children_map[parent] = []
        children_map[parent].append(item)

    # Sort children by their order in original data
    for parent in children_map:
        children_map[parent].sort(key=lambda x: data.index(x))

    # Track Kring names for comments
    kring_counter = [0]  # Use list to allow mutation in nested function
    current_kring_name = [None]

    def get_next_kring_name():
        name = ""
        n = kring_counter[0]
        while True:
            name = chr(ord('A') + n % 26) + name
            n = n // 26 - 1
            if n < 0:
                break
        kring_counter[0] += 1
        return name

    def build_node(item, child_nr=None, kring_name=None):
        props = item["props"]
        node_type = props["type"]

        # Track Kring names
        if node_type == "Kring":
            current_kring_name[0] = get_next_kring_name()

        # Strip default values
        non_default_props = strip_defaults(node_type, props)

        # Build children recursively
        children = []
        is_kring = node_type == "Kring"
        child_counter = 1
        for child in children_map.get(item["id"], []):
            if is_kring:
                children.append(build_node(child, child_nr=child_counter, kring_name=current_kring_name[0]))
                child_counter += 1
            else:
                children.append(build_node(child, kring_name=kring_name))

        # Build comment for auto-generated refs
        comment = None
        if node_type == "Kring" and current_kring_name[0]:
            comment = f"#{current_kring_name[0]}"
        elif kring_name and child_nr:
            comment = f"#{kring_name}.{child_nr}"

        # Build YAML node - put _comment first so it appears right after the type key
        result = {}
        if comment:
            result["_comment"] = comment
        if non_default_props:
            result.update(non_default_props)
        if children:
            result["children"] = children

        # Use type as key if we have props or children
        if result:
            return {node_type: result}
        else:
            return node_type  # Simple string form

    # Build roots (parent = 0)
    roots = []
    for item in children_map.get(0, []):
        roots.append(build_node(item))

    return roots

def build_id_to_ref(data):
    """Build mapping from element id to ref string (KringName.nr)."""
    id_to_ref = {}
    kring_counter = 0
    current_kring_name = None

    def get_next_kring_name():
        nonlocal kring_counter
        name = ""
        n = kring_counter
        while True:
            name = chr(ord('A') + n % 26) + name
            n = n // 26 - 1
            if n < 0:
                break
        kring_counter += 1
        return name

    for item in data:
        props = item["props"]
        node_type = props.get("type")

        if node_type == "Kring":
            current_kring_name = get_next_kring_name()

        if current_kring_name and props.get("nr"):
            ref = f"{current_kring_name}.{props['nr']}"
            id_to_ref[item["id"]] = ref

    return id_to_ref

def extract_image_from_svg(svg_text, output_dir, index):
    """Extract embedded image from SVG and save to file. Returns filename."""
    import re
    import base64

    # Look for embedded base64 image: data:image/png;base64,xxxxx
    match = re.search(r'data:image/(png|jpeg|jpg);base64,([A-Za-z0-9+/=]+)', svg_text)
    if not match:
        return None

    img_format = match.group(1)
    if img_format == 'jpeg':
        img_format = 'jpg'
    b64_data = match.group(2)

    filename = f"image_{index}.{img_format}"
    filepath = os.path.join(output_dir, filename)

    with open(filepath, 'wb') as f:
        f.write(base64.b64decode(b64_data))

    return filename

def build_sitplan_yaml(sitplanjson, id_to_ref, output_dir=None):
    """Convert sitplanjson to YAML sitplan format."""
    if not sitplanjson or not sitplanjson.get("elements"):
        return None

    defaults = sitplanjson.get("defaults", {})
    default_scale = defaults.get("scale", 0.7)
    default_rotate = defaults.get("rotate", 0)
    default_fontsize = defaults.get("fontsize", 10)

    # Use 'layers' to preserve order (images and elements interleaved)
    layers = []
    image_index = 0

    for elem in sitplanjson.get("elements", []):
        electro_id = elem.get("electroItemId")
        svg_data = elem.get("svg", "")

        if electro_id is None and svg_data:
            # This is an embedded image layer
            yaml_layer = {
                "pos": FlowList([round(elem["posx"], 1), round(elem["posy"], 1)]),
                "size": FlowList([elem.get("sizex", 0), elem.get("sizey", 0)])
            }

            # Extract and save image if output_dir provided
            if output_dir:
                filename = extract_image_from_svg(svg_data, output_dir, image_index)
                if filename:
                    yaml_layer["image"] = filename
                    image_index += 1
            else:
                # Keep inline SVG reference
                yaml_layer["svg_length"] = len(svg_data)  # For debugging

            if elem.get("scale", default_scale) != default_scale:
                yaml_layer["scale"] = elem["scale"]
            if elem.get("rotate", default_rotate) != default_rotate:
                yaml_layer["rotate"] = elem["rotate"]
            if elem.get("page", 1) != 1:
                yaml_layer["page"] = elem["page"]

            layers.append(yaml_layer)
        elif electro_id is not None:
            # This is an electrical element reference
            ref = id_to_ref.get(electro_id)

            yaml_layer = {
                "pos": FlowList([round(elem["posx"], 1), round(elem["posy"], 1)])
            }

            # Use human-readable ref if available, otherwise use internal id
            if ref:
                yaml_layer["ref"] = ref
            else:
                yaml_layer["id"] = electro_id

            # Only include non-default values
            if elem.get("page", 1) != 1:
                yaml_layer["page"] = elem["page"]
            if elem.get("scale") != default_scale:
                yaml_layer["scale"] = elem["scale"]
            if elem.get("rotate") != default_rotate:
                yaml_layer["rotate"] = elem["rotate"]
            if elem.get("color", "#000000") not in ("#000000", "black"):
                yaml_layer["color"] = elem["color"]

            layers.append(yaml_layer)

    if not layers:
        return None

    result = {
        "defaults": {
            "fontsize": default_fontsize,
            "scale": default_scale,
            "rotate": default_rotate
        },
        "layers": layers
    }

    return result

def eds_to_yaml_data(eds_path, output_dir=None):
    """Convert EDS file to YAML data structure.

    Args:
        eds_path: Path to EDS file
        output_dir: Directory to save extracted images (optional)
    """
    eds_json = eds_to_json(eds_path)

    # Build tree from data
    tree = build_tree(eds_json["data"])

    # Build properties
    props = eds_json.get("properties", {})
    properties = {}
    if props.get("owner"):
        properties["owner"] = props["owner"].replace("<br>", "\n")
    if props.get("installer"):
        properties["installer"] = props["installer"].replace("<br>", "\n")
    if props.get("control"):
        properties["control"] = props["control"].replace("<br>", "\n")
    if props.get("info") and props["info"] != "2 x 230V ~50 Hz":
        properties["info"] = props["info"].replace("<br>", "\n")

    # Build sitplan
    id_to_ref = build_id_to_ref(eds_json["data"])
    sitplan = build_sitplan_yaml(eds_json.get("sitplanjson"), id_to_ref, output_dir)

    # Build result with explicit ordering
    result = {}

    # Add metadata
    result["metadata"] = {
        "schema_version": SCHEMA_VERSION,
        "tool_version": TOOL_VERSION,
        "generated_at": datetime.now(timezone.utc).isoformat()
    }

    if properties:
        result["properties"] = properties
    if sitplan:
        result["sitplan"] = sitplan
    result["elements"] = tree

    return result

def add_yaml_comments(yaml_text):
    """Convert _comment fields to inline YAML comments."""
    import re
    lines = yaml_text.split('\n')
    result = []
    i = 0
    while i < len(lines):
        line = lines[i]
        # Check if next line is a _comment field
        if i + 1 < len(lines):
            next_line = lines[i + 1]
            comment_match = re.match(r'^(\s*)_comment:\s*["\']?(#[A-Z]+(?:\.\d+)?)["\']?\s*$', next_line)
            if comment_match:
                # Add comment to end of current line
                comment = comment_match.group(2)
                # Check if current line ends with `:` or is a type key
                if line.rstrip().endswith(':'):
                    result.append(f"{line} {comment}")
                else:
                    result.append(f"{line}  {comment}")
                i += 2  # Skip the _comment line
                continue
        result.append(line)
        i += 1
    return '\n'.join(result)

def eds_to_yaml(eds_path, yaml_path):
    """Convert EDS file to YAML file.

    Images are extracted to the same directory as the YAML file.
    """
    output_dir = os.path.dirname(os.path.abspath(yaml_path))
    data = eds_to_yaml_data(eds_path, output_dir)

    yaml_text = yaml.dump(data, allow_unicode=True, default_flow_style=False, sort_keys=False)
    yaml_text = add_yaml_comments(yaml_text)

    with open(yaml_path, "w", encoding="utf-8") as f:
        f.write(yaml_text)

    print(f"Converted {eds_path} -> {yaml_path}")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python eds_to_yaml.py input.eds [output.yaml]")
        print("       If output.yaml is omitted, uses input name with .yaml extension")
        sys.exit(1)

    eds_path = sys.argv[1]
    if len(sys.argv) >= 3:
        yaml_path = sys.argv[2]
    else:
        yaml_path = eds_path.rsplit(".", 1)[0] + ".yaml"

    eds_to_yaml(eds_path, yaml_path)

File: tools-ts/src/eds-to-yaml.ts
/**
 * EDS to YAML converter
 *
 * Converts EDS binary format to human-readable YAML format
 */

import * as fs from 'fs';
import * as path from 'path';
import * as YAML from 'yaml';
import { edsToJson, EdsData, EdsItem, SitplanJson } from './eds-codec.js';
import { getDefaults, SCHEMA_VERSION } from './schema.js';

export const TOOL_VERSION = '1.0.0';

// Types for YAML output
interface YamlNode {
  [key: string]: unknown;
}

interface YamlSitplanLayer {
  pos: [number, number];
  ref?: string;
  id?: number;
  scale?: number;
  rotate?: number;
  page?: number;
  color?: string;
  image?: string;
  size?: [number, number];
  svg_length?: number;
}

interface YamlSitplan {
  defaults: {
    fontsize: number;
    scale: number;
    rotate: number;
  };
  layers: YamlSitplanLayer[];
}

interface YamlMetadata {
  schema_version: string;
  tool_version: string;
  generated_at: string;
}

interface YamlDocument {
  metadata?: YamlMetadata;
  properties?: Record<string, string>;
  sitplan?: YamlSitplan;
  elements: YamlNode[];
}

/**
 * Remove properties that match defaults, keeping only non-default values
 */
function stripDefaults(
  nodeType: string,
  props: Record<string, unknown>
): Record<string, unknown> {
  const defaults = getDefaults(nodeType);
  const result: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(props)) {
    // Skip type and nr, handled separately
    if (key === 'type' || key === 'nr') continue;

    // Skip auto-generated Kring names
    if (key === 'naam' && nodeType === 'Kring') continue;

    // Skip auto values
    if ((key === 'autonr' || key === 'autoKringNaam') && value === 'auto') {
      continue;
    }

    const defaultVal = defaults[key];
    // Compare as strings since EDS often stores numbers as strings
    if (defaultVal !== undefined && String(value) === String(defaultVal)) {
      continue;
    }

    result[key] = value;
  }

  return result;
}

/**
 * Generate next Kring name (A, B, C, ... Z, AA, AB, ...)
 */
function createKringNameGenerator(): () => string {
  let counter = 0;

  return () => {
    let name = '';
    let n = counter;
    while (true) {
      name = String.fromCharCode('A'.charCodeAt(0) + (n % 26)) + name;
      n = Math.floor(n / 26) - 1;
      if (n < 0) break;
    }
    counter++;
    return name;
  };
}

/**
 * Build tree structure from flat EDS data
 */
function buildTree(data: EdsItem[]): YamlNode[] {
  // Build id -> item lookup
  const idToItem = new Map<number, EdsItem>();
  for (const item of data) {
    idToItem.set(item.id, item);
  }

  // Build parent -> children lookup
  const childrenMap = new Map<number, EdsItem[]>();
  for (const item of data) {
    const parent = item.parent;
    if (!childrenMap.has(parent)) {
      childrenMap.set(parent, []);
    }
    childrenMap.get(parent)!.push(item);
  }

  // Sort children by their order in original data
  for (const [, children] of childrenMap) {
    children.sort((a, b) => data.indexOf(a) - data.indexOf(b));
  }

  // Track Kring names for comments
  const getNextKringName = createKringNameGenerator();
  let currentKringName: string | null = null;

  function buildNode(
    item: EdsItem,
    childNr?: number,
    kringName?: string | null
  ): YamlNode {
    const props = item.props as Record<string, unknown>;
    const nodeType = props.type as string;

    // Track Kring names
    if (nodeType === 'Kring') {
      currentKringName = getNextKringName();
    }

    // Strip default values
    const nonDefaultProps = stripDefaults(nodeType, props);

    // Build children recursively
    const children: YamlNode[] = [];
    const isKring = nodeType === 'Kring';
    let childCounter = 1;

    for (const child of childrenMap.get(item.id) || []) {
      if (isKring) {
        children.push(buildNode(child, childCounter, currentKringName));
        childCounter++;
      } else {
        children.push(buildNode(child, undefined, kringName));
      }
    }

    // Build comment for auto-generated refs
    let comment: string | undefined;
    if (nodeType === 'Kring' && currentKringName) {
      comment = `#${currentKringName}`;
    } else if (kringName && childNr) {
      comment = `#${kringName}.${childNr}`;
    }

    // Build YAML node
    const result: Record<string, unknown> = {};
    if (comment) {
      result._comment = comment;
    }
    if (Object.keys(nonDefaultProps).length > 0) {
      Object.assign(result, nonDefaultProps);
    }
    if (children.length > 0) {
      result.children = children;
    }

    // Use type as key if we have props or children
    if (Object.keys(result).length > 0) {
      return { [nodeType]: result };
    } else {
      return nodeType as unknown as YamlNode; // Simple string form
    }
  }

  // Build roots (parent = 0)
  const roots: YamlNode[] = [];
  for (const item of childrenMap.get(0) || []) {
    roots.push(buildNode(item));
  }

  return roots;
}

/**
 * Build mapping from element id to ref string (KringName.nr)
 */
function buildIdToRef(data: EdsItem[]): Map<number, string> {
  const idToRef = new Map<number, string>();
  const getNextKringName = createKringNameGenerator();
  let currentKringName: string | null = null;

  for (const item of data) {
    const props = item.props as Record<string, unknown>;
    const nodeType = props.type as string;

    if (nodeType === 'Kring') {
      currentKringName = getNextKringName();
    }

    if (currentKringName && props.nr) {
      const ref = `${currentKringName}.${props.nr}`;
      idToRef.set(item.id, ref);
    }
  }

  return idToRef;
}

/**
 * Extract embedded image from SVG and save to file
 */
function extractImageFromSvg(
  svgText: string,
  outputDir: string,
  index: number
): string | null {
  // Look for embedded base64 image
  const match = svgText.match(
    /data:image\/(png|jpeg|jpg);base64,([A-Za-z0-9+/=]+)/
  );
  if (!match) return null;

  let imgFormat = match[1];
  if (imgFormat === 'jpeg') imgFormat = 'jpg';
  const b64Data = match[2];

  const filename = `image_${index}.${imgFormat}`;
  const filepath = path.join(outputDir, filename);

  fs.writeFileSync(filepath, Buffer.from(b64Data, 'base64'));

  return filename;
}

/**
 * Convert sitplanjson to YAML sitplan format
 */
function buildSitplanYaml(
  sitplanjson: SitplanJson | null,
  idToRef: Map<number, string>,
  outputDir?: string
): YamlSitplan | null {
  if (!sitplanjson || !sitplanjson.elements?.length) {
    return null;
  }

  const defaults = sitplanjson.defaults || {};
  const defaultScale = defaults.scale ?? 0.7;
  const defaultRotate = defaults.rotate ?? 0;
  const defaultFontsize = defaults.fontsize ?? 10;

  const layers: YamlSitplanLayer[] = [];
  let imageIndex = 0;

  for (const elem of sitplanjson.elements) {
    const electroId = elem.electroItemId;
    const svgData = elem.svg || '';

    if (electroId === null && svgData) {
      // This is an embedded image layer
      const yamlLayer: YamlSitplanLayer = {
        pos: [
          Math.round(elem.posx * 10) / 10,
          Math.round(elem.posy * 10) / 10,
        ],
        size: [elem.sizex || 0, elem.sizey || 0],
      };

      // Extract and save image if output_dir provided
      if (outputDir) {
        const filename = extractImageFromSvg(svgData, outputDir, imageIndex);
        if (filename) {
          yamlLayer.image = filename;
          imageIndex++;
        }
      } else {
        yamlLayer.svg_length = svgData.length;
      }

      if ((elem.scale ?? defaultScale) !== defaultScale) {
        yamlLayer.scale = elem.scale;
      }
      if ((elem.rotate ?? defaultRotate) !== defaultRotate) {
        yamlLayer.rotate = elem.rotate;
      }
      if ((elem.page ?? 1) !== 1) {
        yamlLayer.page = elem.page;
      }

      layers.push(yamlLayer);
    } else if (electroId !== null) {
      // This is an electrical element reference
      const ref = idToRef.get(electroId);

      const yamlLayer: YamlSitplanLayer = {
        pos: [
          Math.round(elem.posx * 10) / 10,
          Math.round(elem.posy * 10) / 10,
        ],
      };

      // Use human-readable ref if available
      if (ref) {
        yamlLayer.ref = ref;
      } else {
        yamlLayer.id = electroId;
      }

      if ((elem.page ?? 1) !== 1) {
        yamlLayer.page = elem.page;
      }
      if (elem.scale !== defaultScale) {
        yamlLayer.scale = elem.scale;
      }
      if (elem.rotate !== defaultRotate) {
        yamlLayer.rotate = elem.rotate;
      }
      if (elem.color && elem.color !== '#000000' && elem.color !== 'black') {
        yamlLayer.color = elem.color;
      }

      layers.push(yamlLayer);
    }
  }

  if (layers.length === 0) {
    return null;
  }

  return {
    defaults: {
      fontsize: defaultFontsize,
      scale: defaultScale,
      rotate: defaultRotate,
    },
    layers,
  };
}

/**
 * Convert EDS file to YAML data structure
 */
export function edsToYamlData(
  edsPath: string,
  outputDir?: string
): YamlDocument {
  const edsJson = edsToJson(edsPath);

  // Build tree from data
  const tree = buildTree(edsJson.data);

  // Build properties
  const props = edsJson.properties || {};
  const properties: Record<string, string> = {};
  if (props.owner) {
    properties.owner = props.owner.replace(/<br>/g, '\n');
  }
  if (props.installer) {
    properties.installer = props.installer.replace(/<br>/g, '\n');
  }
  if (props.control) {
    properties.control = props.control.replace(/<br>/g, '\n');
  }
  if (props.info && props.info !== '2 x 230V ~50 Hz') {
    properties.info = props.info.replace(/<br>/g, '\n');
  }

  // Build sitplan
  const idToRef = buildIdToRef(edsJson.data);
  const sitplan = buildSitplanYaml(edsJson.sitplanjson, idToRef, outputDir);

  // Build result with explicit ordering (metadata, properties, sitplan, elements)
  const result: YamlDocument = { elements: tree };

  // Add metadata
  result.metadata = {
    schema_version: SCHEMA_VERSION,
    tool_version: TOOL_VERSION,
    generated_at: new Date().toISOString(),
  };

  if (Object.keys(properties).length > 0) {
    result.properties = properties;
  }
  if (sitplan) {
    result.sitplan = sitplan;
  }

  return result;
}

/**
 * Convert _comment fields to inline YAML comments
 */
function addYamlComments(yamlText: string): string {
  const lines = yamlText.split('\n');
  const result: string[] = [];
  let i = 0;

  while (i < lines.length) {
    const line = lines[i];

    // Check if next line is a _comment field
    if (i + 1 < lines.length) {
      const nextLine = lines[i + 1];
      const commentMatch = nextLine.match(
        /^(\s*)_comment:\s*['"]?(#[A-Z]+(?:\.\d+)?)['"]?\s*$/
      );

      if (commentMatch) {
        const comment = commentMatch[2];
        // Add comment to end of current line
        if (line.trimEnd().endsWith(':')) {
          result.push(`${line} ${comment}`);
        } else {
          result.push(`${line}  ${comment}`);
        }
        i += 2; // Skip the _comment line
        continue;
      }
    }
    result.push(line);
    i++;
  }

  return result.join('\n');
}

/**
 * Serialize YAML with explicit key ordering
 */
function serializeYaml(data: YamlDocument): string {
  const parts: string[] = [];

  // Metadata section
  if (data.metadata) {
    parts.push('metadata:');
    parts.push(`  schema_version: "${data.metadata.schema_version}"`);
    parts.push(`  tool_version: "${data.metadata.tool_version}"`);
    parts.push(`  generated_at: "${data.metadata.generated_at}"`);
    parts.push('');
  }

  // Properties section
  if (data.properties && Object.keys(data.properties).length > 0) {
    parts.push('properties:');
    for (const [key, value] of Object.entries(data.properties)) {
      if (value.includes('\n')) {
        parts.push(`  ${key}: |`);
        for (const line of value.split('\n')) {
          parts.push(`    ${line}`);
        }
      } else {
        parts.push(`  ${key}: ${value}`);
      }
    }
    parts.push('');
  }

  // Sitplan section
  if (data.sitplan) {
    const sitplanDoc = new YAML.Document({ sitplan: data.sitplan });
    let sitplanText = sitplanDoc.toString({ lineWidth: 0 });
    // Convert pos/size arrays to flow style
    sitplanText = sitplanText.replace(
      /pos:\s*\n\s*-\s*(\d+(?:\.\d+)?)\s*\n\s*-\s*(\d+(?:\.\d+)?)/g,
      'pos: [$1, $2]'
    );
    sitplanText = sitplanText.replace(
      /size:\s*\n\s*-\s*(\d+(?:\.\d+)?)\s*\n\s*-\s*(\d+(?:\.\d+)?)/g,
      'size: [$1, $2]'
    );
    parts.push(sitplanText.trim());
    parts.push('');
  }

  // Elements section
  const elementsDoc = new YAML.Document({ elements: data.elements });
  let elementsText = elementsDoc.toString({ lineWidth: 0 });
  parts.push(elementsText.trim());

  return parts.join('\n');
}

/**
 * Convert EDS file to YAML file
 */
export function edsToYaml(edsPath: string, yamlPath: string): void {
  const outputDir = path.dirname(path.resolve(yamlPath));
  const data = edsToYamlData(edsPath, outputDir);

  let yamlText = serializeYaml(data);
  yamlText = addYamlComments(yamlText);

  fs.writeFileSync(yamlPath, yamlText, 'utf-8');
  console.log(`Converted ${edsPath} -> ${yamlPath}`);
}

export { buildTree, stripDefaults, buildIdToRef };
